blockchain/blockdao/blob_store.go

```diff
@@ -22,4 +40,228 @@ type (
 		GetBlobsByHeight(uint64) ([]*types.BlobTxSidecar, []string, error)
 		PutBlock(*block.Block) error
 	}
+
+	// storage for past N-day's blobs, structured as blow:
+	// 1. Block height is used as key to store blobs. When a new blob is stored,
+	//    the expired time (current - 18 days) is updated and all blobs earlier
+	//    than this time will be deleted.
+	// 2. Two mappings are kept in a separate index: (1) height to tx hashes in,
+	//    the blob, and (2) tx hash to the blob's height. Here's an example:
+	//    304 --> { []hash: eddf9e61fb9d8f5111840daef55e5fde
+	//                      95977fb340867fa13f787df864ed5388 }
+	//    and:
+	//    eddf9e61fb9d8f5111840daef55e5fde --> {height: 311344}
+	//    95977fb340867fa13f787df864ed5388 --> {height: 311344}
+	//    The 2nd mapping is used to support GetBlob(hash) function. It converts
+	//    the requested blob hash to the height, and retrieves the blob.
+	//    When new blobs are added to the storage, the index of expired blobs will
+	//    be deleted correspondingly.
+	// 3. The maximum number of blobs is 6 for each block, so maximum data size
+	//    stored by a key is 131kB x 6 = 786kB. The maximum total size of the
+	//    entire blob storage is 786kB x 311040 = 245GB.
+	//
+	blobStore struct {
+		kvStore                         db.KVStore
+		totalBlocks                     uint64
+		currWriteBlock, currExpireBlock uint64
+		purgeTask                       *routine.RecurringTask
+		interval                        time.Duration
+	}
 )
+
+func NewBlobStore(kv db.KVStore, size uint64, interval time.Duration) *blobStore {
+	bs := blobStore{
+		kvStore:     kv,
+		totalBlocks: size,
+		interval:    interval,
+	}
+	return &bs
+}
+
+func (bs *blobStore) Start(ctx context.Context) error {
+	if err := bs.kvStore.Start(ctx); err != nil {
+		return err
+	}
+	if err := bs.checkDB(); err != nil {
+		return err
+	}
+	if bs.interval != 0 {
+		bs.purgeTask = routine.NewRecurringTask(func() {
+			bs.expireBlob(bs.currWriteBlock)
```

Comment:
`currWriteBlock` should be protected for data race

Reply:
It might be more appropriate to place the purgeTask creation in the NewBlobStore, but it doesn't matter much.